ggml/src/ggml-qnn/backend-ops.cpp

```diff
@@ -0,0 +1,542 @@
+            return dst->src[0];
+        case 2:
+            return dst->src[0] && dst->src[1];
+        default:
+            QNN_LOG_WARN("invalid op param count %d", (int)param_count);
+            break;
+    }
+
+    return false;
+}
+
+#ifndef NDEBUG
+void print_ggml_tensor(const ggml_tensor *tensor) {
+    QNN_LOG_DEBUG("%s: type:%s ne: %ldx%ldx%ldx%ld, nb: %ldx%ldx%ldx%ld", tensor->name, ggml_type_name(tensor->type),
+                  (long)tensor->ne[0], (long)tensor->ne[1], (long)tensor->ne[2], (long)tensor->ne[3],
+                  (long)tensor->nb[0], (long)tensor->nb[1], (long)tensor->nb[2], (long)tensor->nb[3]);
+}
+#endif
+
+} // namespace
+
+namespace {
+
+typedef bool (*ggml_qnn_op_t)(ggml_backend_qnn_device_context *ctx, ggml_tensor *dst);
+
+bool execute_graph(qnn::qnn_graph *graph, ggml_tensor *output) {
+    if (!graph->execute(output)) {
+        QNN_LOG_WARN("execute failed");
+        return false;
+    }
+
+    return true;
+}
+
+void append_tensor_dimensions(const ggml_tensor *tensor, std::string &output) {
+    char buffer[256] = {};
+    const auto *type_name = qnn::get_ggml_type_name(tensor->type);
+    int len = 0;
+    switch (ggml_n_dims(tensor)) {
+        case 1:
+            len = snprintf(buffer, sizeof(buffer), "%ld%s", (long)tensor->ne[0], type_name);
+            break;
+        case 2:
+            len = snprintf(buffer, sizeof(buffer), "%ldx%ld%s", (long)tensor->ne[0], (long)tensor->ne[1], type_name);
+            break;
+        case 3:
+            len = snprintf(buffer, sizeof(buffer), "%ldx%ldx%ld%s", (long)tensor->ne[0], (long)tensor->ne[1],
+                           (long)tensor->ne[2], type_name);
+            break;
+        case 4:
+        default:
+            len = snprintf(buffer, sizeof(buffer), "%ldx%ldx%ldx%ld%s", (long)tensor->ne[0], (long)tensor->ne[1],
+                           (long)tensor->ne[2], (long)tensor->ne[3], type_name);
+            break;
+    }
+    GGML_ASSERT(len > 0 && len < (int)sizeof(buffer));
+    output.append(buffer, len);
+}
+
+void get_graph_key_from_op(const ggml_tensor *op, std::string &output) {
+    GGML_ASSERT(op->op != GGML_OP_NONE);
+    output += ggml_op_desc(op);
+    output += qnn::get_ggml_type_name(op->type);
+    const auto param_count = qnn::get_qnn_op_input_param_count(op);
+    for (size_t i = 0; i < param_count; ++i) {
+        auto *input = op->src[i];
+        if (!input) {
+            break;
+        }
+
+        output += '_';
+        append_tensor_dimensions(input, output);
+    }
+}
+
+void get_op_key_with_src_op_desc(const ggml_tensor *op, std::string &output) {
+    output += ggml_op_desc(op);
+    output += '(';
+    if (op->src[0]) {
+        output += ggml_op_desc(op->src[0]);
+    }
+    for (size_t i = 1; i < GGML_MAX_DIMS && op->src[i]; ++i) {
+        output += ',';
+        output += ggml_op_desc(op->src[i]);
+    }
+    output += ')';
+}
+
+void get_graph_key_from_cgraph(const ggml_cgraph *cgraph, std::string &output) {
```

Comment:
Generates a unique key for a given `ggml_cgraph`. The key is constructed by concatenating the descriptions of the operations and their associated tensor dimensions within the graph.

Example key format: `MUL_MATf32_256x16x10f32_256x1x10f32#LOG#ADD#ADDf32_16x1x10f32`

May need some refactoring here to handle more complex graph structures and edge cases